pkg/cloudops/aws/aws.go

```diff
@@ -0,0 +1,257 @@
+					Region: a.instInfo.Region,
+					Labels: labelsFromTags(group.Tags),
+				},
+				Zones:              zones,
+				AutoscalingEnabled: true,
+				Min:                *group.MinSize,
+				Max:                *group.MaxSize,
+			}
+
+			return retval, nil
+		}
+	}
+
+	return nil, fmt.Errorf("instance doesn't belong to an instance group")
+}
+
+func (a *awsOps) populateBasicInfo() error {
+	var err error
+	a.instInfo.Zone, a.instInfo.ID, err = getInfoFromMetadata()
+	if err != nil {
+		logrus.Infof("Failed to query ec2 metadata info. Instance may not be running on ec2.")
+	} else {
+		return nil
+	}
+
+	// try env variables
+	a.instInfo.Zone, err = util.GetEnvValueStrict("AWS_ZONE")
+	if err != nil {
+		return err
+	}
+
+	a.instInfo.ID, err = util.GetEnvValueStrict("AWS_INSTANCE_ID")
+	if err != nil {
+		return err
+	}
+
+	if _, err := credentials.NewEnvCredentials().Get(); err != nil {
+		return ErrAWSEnvNotAvailable
+	}
+
+	return nil
+
+}
+
+func getInfoFromMetadata() (string, string, error) {
+	zone, err := metadata("placement/availability-zone")
+	if err != nil {
+		return "", "", err
+	}
+
+	instanceID, err := metadata("instance-id")
+	if err != nil {
+		return "", "", err
+	}
+
+	return zone, instanceID, nil
+}
+
+// instanceData retrieves instance data specified by key.
+func instanceData(key string) (string, error) {
+	client := http.Client{Timeout: time.Second * 3}
+	url := "http://169.254.169.254/latest/" + key
+	res, err := client.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer res.Body.Close()
+	if res.StatusCode != 200 {
+		err = fmt.Errorf("Code %d returned for url %s", res.StatusCode, url)
+		return "", fmt.Errorf("Error querying AWS metadata for key %s: %v", key, err)
+	}
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return "", fmt.Errorf("Error querying AWS metadata for key %s: %v", key, err)
+	}
+	if len(body) == 0 {
+		return "", fmt.Errorf("Failed to retrieve AWS metadata for key %s: %v", key, err)
+	}
+	return string(body), nil
+}
+
+func metadata(key string) (string, error) {
+	return instanceData("meta-data/" + key)
+}
+
+// DescribeInstanceByID describes the given instance by instance ID
+func DescribeInstanceByID(service *ec2.EC2, id string) (*ec2.Instance, error) {
+	request := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{&id},
+	}
+	out, err := service.DescribeInstances(request)
+	if err != nil {
+		return nil, err
+	}
+	if len(out.Reservations) != 1 {
+		return nil, fmt.Errorf("DescribeInstances(%v) returned %v reservations, expect 1",
+			id, len(out.Reservations))
+	}
+	if len(out.Reservations[0].Instances) != 1 {
+		return nil, fmt.Errorf("DescribeInstances(%v) returned %v Reservations, expect 1",
+			id, len(out.Reservations[0].Instances))
+	}
+	return out.Reservations[0].Instances[0], nil
+}
+
+func labelsFromTags(input interface{}) map[string]string {
+	labels := make(map[string]string)
+	ec2Tags, ok := input.([]*ec2.Tag)
+	if ok {
+		for _, tag := range ec2Tags {
```

Comment:
nit: you can have a function to loop over tags, to reduce code duplication

Reply:
the tag types are different (ec2.Tag vs autoscaling.TagDescriptio). Hence the duplicate code to process them.